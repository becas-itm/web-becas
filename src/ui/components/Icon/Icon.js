import React from 'react';
import propTypes from 'prop-types';
import classNames from 'classnames';

import './Icon.scss';

export const SIZE = {
  small: '-sizeSmall',
  regular: '-sizeRegular',
  auto: '-sizeAuto',
};

export function Icon({ size, title, className, children, ...props }) {
  const iconClasses = classNames('Icon', size, className);

  const a11yProps = title ? { role: 'img' } : { 'aria-hidden': true };

  return (
    <svg {...a11yProps} {...props} className={iconClasses}>
      {title && <title>{title}</title>}
      {children}
    </svg>
  );
}

Icon.defaultProps = {
  title: null,
  size: SIZE.regular,
};

Icon.propTypes = {
  title: propTypes.string,
  children: propTypes.node,
  className: propTypes.string,
  size: propTypes.oneOf(Object.values(SIZE)),
};
