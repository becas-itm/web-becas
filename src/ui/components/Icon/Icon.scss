.Icon {
  display: inline-block;
  color: currentColor;
  fill: currentColor;

  &.-sizeSmall {
    width: 18px;
    height: 18px;
  }

  &.-sizeRegular {
    width: 24px;
    height: 24px;
  }

  &.-sizeAuto {
    width: 100%;
    height: 100%;
  }
}
