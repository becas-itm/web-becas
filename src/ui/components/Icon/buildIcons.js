const fs = require('fs');
const path = require('path');

const SVGO = require('svgo');
const prettier = require('prettier');

const svgDir = path.join(__dirname, 'svg');
const iconsDir = path.join(__dirname, 'icons');

const templateFile = './iconTemplate.txt';

const pascalCase = string => string.charAt(0).toUpperCase() + string.substr(1);

const svgo = new SVGO({
  plugins: [
    { removeViewBox: false },
    { removeDimensions: true },
    {
      removeAttrs: { attrs: '(xmlns)' },
    },
  ],
});

function clearOldIcons() {
  fs.readdirSync(iconsDir)
    .filter(file => /\.js$/.test(file))
    .map(file => path.join(iconsDir, file))
    .forEach(filePath => fs.unlinkSync(filePath));
}

function reactify(svgString) {
  return svgString
    .replace('<svg ', '<Icon {...props} ')
    .replace('</svg>', '</Icon>');
}

async function generateIcons() {
  const iconTemplate = fs.readFileSync(
    path.resolve(__dirname, templateFile),
    'utf8',
  );

  const svgs = fs.readdirSync(svgDir).filter(file => file.endsWith('.svg'));

  const prettierOptions = (await prettier.resolveConfig(__dirname)) || {};
  const iconExports = [];

  svgs.forEach(async svgFilename => {
    const componentName = pascalCase(path.parse(svgFilename).name);
    iconExports.push(`export { ${componentName} } from './${componentName}';`);

    const svgContent = fs.readFileSync(
      path.resolve(svgDir, svgFilename),
      'utf8',
    );

    const data = await svgo
      .optimize(svgContent)
      .then(result => result.data)
      .then(result =>
        iconTemplate
          .replace(new RegExp('%%ICON_NAME%%', 'g'), componentName)
          .replace('%%ICON_SVG%%', reactify(result)),
      );

    fs.writeFileSync(
      path.resolve(iconsDir, `./${componentName}.js`),
      prettier.format(data, {
        ...prettierOptions,
        parser: 'babel',
      }),
    );
  });

  fs.writeFileSync(
    path.resolve(iconsDir, './index.js'),
    `${iconExports.join('\n\n')}\n`,
  );

  // eslint-disable-next-line no-console
  console.log(`Wrote ${svgs.length} icon(s)`);
}

clearOldIcons();

generateIcons();
