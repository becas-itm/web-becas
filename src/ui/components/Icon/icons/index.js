export { ArrowBack } from './ArrowBack';

export { ArrowForward } from './ArrowForward';

export { Close } from './Close';

export { Event } from './Event';

export { Search } from './Search';

export { Tune } from './Tune';
