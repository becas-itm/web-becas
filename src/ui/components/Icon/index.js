export * from './icons';

export { default, default as Icon } from './Icon';
