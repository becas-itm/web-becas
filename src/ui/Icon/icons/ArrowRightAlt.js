import React from 'react';
import Icon from '../Icon';

export function ArrowRightAlt(props) {
  return (
    <Icon {...props} viewBox="0 0 24 24">
      <path d="M0 0h24v24H0z" fill="none" />
      <path d="M16.01 11H4v2h12.01v3L20 12l-3.99-4z" />
    </Icon>
  );
}

ArrowRightAlt.defaultProps = Icon.defaultProps;

ArrowRightAlt.propTypes = Icon.propTypes;
