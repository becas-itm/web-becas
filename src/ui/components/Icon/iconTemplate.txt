import React from 'react';
import Icon from '../Icon';

export function %%ICON_NAME%%(props) {
  return %%ICON_SVG%%;
}

%%ICON_NAME%%.defaultProps = Icon.defaultProps;

%%ICON_NAME%%.propTypes = Icon.propTypes;
