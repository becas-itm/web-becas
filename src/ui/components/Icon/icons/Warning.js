import React from 'react';
import Icon from '../Icon';

export function Warning(props) {
  return (
    <Icon {...props} viewBox="0 0 24 24">
      <path d="M0 0h24v24H0z" fill="none" />
      <path d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z" />
    </Icon>
  );
}

Warning.defaultProps = Icon.defaultProps;

Warning.propTypes = Icon.propTypes;
