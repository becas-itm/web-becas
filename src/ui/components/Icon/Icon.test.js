import React from 'react';
import { render } from '@testing-library/react';

import Icon from './index';
import * as icons from './icons';

const Content = () => <div data-testid="icon-content" />;

const TestIcon = props => (
  <Icon data-testid="Icon" {...props}>
    <Content />
  </Icon>
);

describe('Icon component', () => {
  test('renders correctly', () => {
    const { getByTestId } = render(<TestIcon />);
    expect(getByTestId('icon-content')).toBeInTheDocument();
  });

  describe('icon title', () => {
    it('null title should have aria-hidden attribute', () => {
      const { getByTestId } = render(<TestIcon />);
      expect(getByTestId('Icon')).toHaveAttribute('aria-hidden', 'true');
    });

    it('role attribute should be `img`', () => {
      const title = 'foo';
      const { getByTestId } = render(<TestIcon title={title} />);
      const icon = getByTestId('Icon');
      expect(icon).toHaveAttribute('role', 'img');
      expect(icon).not.toHaveAttribute('aria-hidden');
    });

    it('should render a title element', () => {
      const title = 'foo';
      const { getByTitle } = render(<TestIcon title={title} />);
      expect(getByTitle(title)).toHaveTextContent(title);
    });
  });

  describe('all icons', () => {
    test.each(Object.entries(icons))(
      '%s icon renders correctly',
      (icon, IconComp) => {
        const { getByTestId } = render(<IconComp data-testid={icon} />);
        expect(getByTestId(icon)).toMatchSnapshot();
      },
    );
  });
});
