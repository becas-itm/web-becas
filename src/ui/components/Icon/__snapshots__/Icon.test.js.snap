// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component all icons Add icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Add"
  viewBox="0 0 24 24"
>
  <path
    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons ArrowBack icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ArrowBack"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
  />
</svg>
`;

exports[`Icon component all icons ArrowForward icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ArrowForward"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
  />
</svg>
`;

exports[`Icon component all icons ArrowRightAlt icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ArrowRightAlt"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M16.01 11H4v2h12.01v3L20 12l-3.99-4z"
  />
</svg>
`;

exports[`Icon component all icons ChevronLeft icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ChevronLeft"
  viewBox="0 0 24 24"
>
  <path
    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons ChevronRight icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ChevronRight"
  viewBox="0 0 24 24"
>
  <path
    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Close icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Close"
  viewBox="0 0 24 24"
>
  <path
    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Edit icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Edit"
  viewBox="0 0 24 24"
>
  <path
    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Event icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Event"
  viewBox="0 0 24 24"
>
  <path
    d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons ExitToApp icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="ExitToApp"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
  />
</svg>
`;

exports[`Icon component all icons FirstPage icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="FirstPage"
  viewBox="0 0 24 24"
>
  <path
    d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
  />
  <path
    d="M24 24H0V0h24v24z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Github icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Github"
  viewBox="0 0 24 24"
>
  <path
    d="M12 2A10 10 0 002 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0012 2z"
  />
</svg>
`;

exports[`Icon component all icons Home icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Home"
  viewBox="0 0 24 24"
>
  <path
    d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Inbox icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Inbox"
  viewBox="0 0 24 24"
>
  <path
    d="M19 3H4.99c-1.11 0-1.98.89-1.98 2L3 19c0 1.1.88 2 1.99 2H19c1.1 0 2-.9 2-2V5a2 2 0 00-2-2zm0 12h-4c0 1.66-1.35 3-3 3s-3-1.34-3-3H4.99V5H19v10z"
  />
  <path
    d="M0 0h24v24H0V0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons KeyboardArrowDown icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="KeyboardArrowDown"
  viewBox="0 0 24 24"
>
  <path
    d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
  />
  <path
    d="M0 0h24v24H0V0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons LastPage icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="LastPage"
  viewBox="0 0 24 24"
>
  <path
    d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
  />
  <path
    d="M0 0h24v24H0V0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons Menu icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Menu"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
  />
</svg>
`;

exports[`Icon component all icons Money icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Money"
  viewBox="0 0 24 24"
>
  <path
    d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons OpenInNew icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="OpenInNew"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M19 19H5V5h7V3H5a2 2 0 00-2 2v14a2 2 0 002 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
  />
</svg>
`;

exports[`Icon component all icons School icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="School"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M5 13.18v4L12 21l7-3.82v-4L12 17l-7-3.82zM12 3L1 9l11 6 9-4.91V17h2V9L12 3z"
  />
</svg>
`;

exports[`Icon component all icons Search icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Search"
  viewBox="0 0 24 24"
>
  <path
    d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`Icon component all icons SettingsBackupRestore icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="SettingsBackupRestore"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M14 12c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-2-9a9 9 0 00-9 9H0l4 4 4-4H5c0-3.87 3.13-7 7-7s7 3.13 7 7a6.995 6.995 0 01-11.06 5.7l-1.42 1.44A9 9 0 1012 3z"
  />
</svg>
`;

exports[`Icon component all icons SupervisorAccount icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="SupervisorAccount"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M16.5 12c1.38 0 2.49-1.12 2.49-2.5S17.88 7 16.5 7a2.5 2.5 0 000 5zM9 11c1.66 0 2.99-1.34 2.99-3S10.66 5 9 5C7.34 5 6 6.34 6 8s1.34 3 3 3zm7.5 3c-1.83 0-5.5.92-5.5 2.75V19h11v-2.25c0-1.83-3.67-2.75-5.5-2.75zM9 13c-2.33 0-7 1.17-7 3.5V19h7v-2.25c0-.85.33-2.34 2.37-3.47C10.5 13.1 9.66 13 9 13z"
  />
</svg>
`;

exports[`Icon component all icons Tune icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Tune"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M3 17v2h6v-2H3zM3 5v2h10V5H3zm10 16v-2h8v-2h-8v-2h-2v6h2zM7 9v2H3v2h4v2h2V9H7zm14 4v-2H11v2h10zm-6-4h2V7h4V5h-4V3h-2v6z"
  />
</svg>
`;

exports[`Icon component all icons Warning icon renders correctly 1`] = `
<svg
  aria-hidden="true"
  class="inline-block fill-current w-6 h-6"
  data-testid="Warning"
  viewBox="0 0 24 24"
>
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
  <path
    d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
  />
</svg>
`;
